File: GARI_Umbria_2026-08-03/DotNet/AgronicaCoreStd/AgronicaCoreModelsSTD/anagrafiche/ParcoMacchine.cs
using AgronicaCoreModelsSTD.baseClass;
using AgronicaCoreModelsSTD.documenti;
using AgronicaCoreModelsSTD.metaschema;
using System;
using System.Collections.Generic;
using System.Text;

namespace AgronicaCoreModelsSTD.anagrafiche
{
    public class ParcoMacchine
    {
        public string partitaIva { get; set; }
        public CentroAziendale.PK centroPK { get; set; }

        public Contatto contatto { get; set; }

        public string descrizione { get; set; }
        public DittaMacchina marca { get; set; }
        public string modello { get; set; }
        public FinalitaMacchina finalita { get; set; }

        /// <summary>
        /// class_cod su database, arriva da tabella macchine
        /// </summary>
        public Macchine tipo { get; set; }

        public MacchineDettaglio1 dettaglio_1 { get; set; }
        public MacchineDettaglio2 dettaglio_2 { get; set; }

        public int codice { get; set; }

        public int codiceOrigine { get; set; }

        public string codice_stringa { get; set; }

        public bool visibilitaPubblica { get; set; }
        public bool? visibileControlloGestione { get; set; }

        public IntervalloTemporale validita { get; set; }

        public DateTime Data_Carico { get; set; }
        public DateTime Data_Scarico { get; set; }

        public string Username_Creazione { get; set; }
        public string Username_Modifica { get; set; }

        public TitoloDiPossesso titolo_Possesso { get; set; }
        public string proprietario { get; set; }
        public string CUAA_Proprietario { get; set; }

        /// Dati Tecnici
        public string targa { get; set; }

        public TipoTarga tipo_Targa { get; set; }

        public string telaio { get; set; }

        public string n_Immatricolazione { get; set; }
        public DateTime data_Immatricolazione { get; set; }
        public string n_Immatricolazione_Rimorchio { get; set; }
        public string N_Autorizzazione_Trasporto { get; set; }
        public DateTime data_Rilascio_Autorizzazione { get; set; }

        public DateTime data_Ultima_Revisione { get; set; }
        public DateTime data_Ultima_Manutenzione { get; set; }
        public Carburante alimentazione { get; set; }
        public AgronicaCoreModelsSTD.Utility.Immagine immagineGrande { get; set; }
        public AgronicaCoreModelsSTD.Utility.Immagine immaginePiccola { get; set; }

        public String VIN { get; set; }
        public String BTM_Serial { get; set; }

        public String ExternalAPIKey { get; set; }

        public BaseCodeDescr HubIoT_PlatformDestination { get; set; }

        public MacchinaGerarchia gerarchiaPadre { get; set; }

        public List<MacchinaGerarchia> gerarchiaFigli { get; set; }

        // 'Taratura Ugello
        public decimal taratura_Ugello { get; set; }
        public string numero_certificato { get; set; }
        public DateTime data_Ultima_Taratura { get; set; }
        public DateTime scadenza_Taratura { get; set; }

        public string stato_Utilizzo { get; set; }
        public string potenza { get; set; }
        public UnitaDiMisura unita_Misura { get; set; }
        public string note { get; set; }
        public bool flag_cancellazione { get; set; }

        public int portata { get; set; }
        public double efficienza { get; set; }
        public int codice_impianto { get; set; }

        public MacchineCodificaAgea ageaCod { get; set; }

        public ParcoMacchine()
        {
            flag_cancellazione = false;
            Documenti = new List<DocumentoPerMacchina>();
        }

        public List<ParcoMacchineCaratteristiche> caratteristiche { get; set; }

        public List<CostoUnitario> costi { get; set; }

        public List<RateoTempo> rateiTempo { get; set; }

        public string Distinta_Installazione { get; set; }

        public string Contratto_Installazione { get; set; }

        public string Tipologia_Installazione { get; set; }
        public DateTime? Data_Inizio_Installazione { get; set; }
        public DateTime? Data_Fine_Installazione { get; set; }
        public string Stato_Installazione { get; set; }
        public string Provincia_Istat_Installazione { get; set; }
        public string Comune_Istat_Installazione { get; set; }
        public string Indirizzo_Installazione { get; set; }
        public double Latitudine_Installazione { get; set; }
        public double Longitudine_Installazione { get; set; }

        //documenti Allegati valorizzati solo da APP
        public List<DocumentoPerMacchina> Documenti { get; set; }

        public class PK
        {
            public int codice { get; set; }
            public CentroAziendale.PK centroAziendalePK { get; set; }

            public PK(int codice, CentroAziendale.PK centroAziendalePK)
            {
                this.codice = codice;
                this.centroAziendalePK = centroAziendalePK;
            }

            public PK(int codice, int saCod, string piva)
            {
                this.codice = codice;
                this.centroAziendalePK = new CentroAziendale.PK(saCod, piva);
            }

            public PK()
            {
            }
        }
    }

    public class MacchinaGerarchia
    {
        public int ID { get; set; }
        public ParcoMacchine macchina { get; set; }
        public BaseCodeDescr legame { get; set; }
        public String desclegame { get; set; }
        public UnitaDiMisura udm { get; set; }
        public int qta { get; set; }
        public DateTime validita_inizio { get; set; }
        public DateTime validita_fine { get; set; }
    }

    public class LinkedMachine<T> : ParcoMacchine
    {
        public IntervalloTemporale linkValidity { get; set; }
        public T linkedItemPK { get; set; }

        public LinkedMachine() : base()
        {

        }
    }
}

File: GARI_Umbria_2026-08-03/GiasNG/src/app/Model/anagrafiche/ParcoMacchine.ts
import { CentroAziendale, PKCentroAziendale } from './CentroAziendale';
import { Macchine } from '../metaschema/Macchine';
import { IntervalloTemporale } from './IntervalloTemporale';
import { UnitaDiMisura} from '../metaschema/UnitaDiMisura';
import { CostoUnitario } from './CostoUnitario';
import { DittaMacchina } from '../metaschema/DittaMacchina';
import { FinalitaMacchina } from '../metaschema/FinalitaMacchina';
import { MacchineDettaglio1 } from '../metaschema/MacchineDettaglio1';
import { MacchineDettaglio2 } from '../metaschema/MacchineDettaglio2';
import { TitoloDiPossesso } from '../metaschema/TitoloDiPossesso';
import { TipoTarga } from '../metaschema/TipoTarga';
import { Carburante } from '../metaschema/Carburante';
import { Immagine } from './Immagine';
import { List } from 'lodash';
import { BaseCodeDescr } from '../baseClass/baseCodeDescr';
import { Tipo } from '../MetaschemaModel';
import { Contatto } from './Contatto';
import {AGRODATAFINE, AGRODATAINIZIO} from '../CostantiPersonalizzate';
import {MacchineCodificaAgea} from '../metaschema/CodificheAgea';
import { D } from '@angular/cdk/bidi-module.d-D-fEBKdS';

export class ParcoMacchine {
  partitaIva: string;
  codice: number;
  centroPK: PKCentroAziendale;

  contatto: Contatto;
  alimentazione: Carburante;
  codice_stringa: string;
  costi: CostoUnitario[];
  CUAA_Proprietario: string;

  data_Immatricolazione: Date;
  Data_Carico: Date;
  Data_Scarico: Date;
  data_Rilascio_Autorizzazione: Date;
  data_Ultima_Revisione: Date;
  data_Ultima_Manutenzione: Date;
  data_Ultima_Taratura: Date;
  descrizione: string;
  dettaglio_1: MacchineDettaglio1;
  dettaglio_2: MacchineDettaglio2;
  ageaCod: MacchineCodificaAgea;

  finalita: FinalitaMacchina;

  marca: DittaMacchina;
  modello: string;

  N_Autorizzazione_Trasporto: string;
  n_Immatricolazione: string;
  n_Immatricolazione_Rimorchio: string;
  note: string;

  potenza: string;
  proprietario: string;

  scadenza_Taratura: Date;
  stato_Utilizzo: string;

  immagineGrande: Immagine;
  immaginePiccola: Immagine;

  VIN: String;
  BTM_Serial: String;
  ExternalAPIKey: String;

  HubIoT_PlatformDestination: BaseCodeDescr;

  caratteristiche: Caratteristica[];

  gerarchiaPadre: MacchinaGerarchia;

  gerarchiaFigli: MacchinaGerarchia[];

  rateiTempo: RateoTempo[];

  taratura_Ugello: number;
  numero_certificato: string;
  targa: string;
  telaio: string;
  tipo: Macchine;
  tipo_Targa: TipoTarga;
  titolo_Possesso: TitoloDiPossesso;

  unita_Misura: UnitaDiMisura;

  validita: IntervalloTemporale;
  visibilitaPubblica: boolean|number;
  visibileControlloGestione: boolean;

  flag_cancellazione: boolean;

  efficienza: number = 0;
  portata: number = 0;
  codice_impianto: number = 0;

  //stazioni meteo
  Distinta_Installazione: string = "";
  Contratto_Installazione: string = "";
  Tipologia_Installazione: string = "";
  Data_Inizio_Installazione: Date;
  Data_Fine_Installazione: Date;
  Stato_Installazione: string = "";
  Provincia_Istat_Installazione: string = "";
  Comune_Istat_Installazione: string = "";
  Indirizzo_Installazione: string = "";
  Latitudine_Installazione: number;
  Longitudine_Installazione: number;


  constructor(macCod: number = 0) {
    this.flag_cancellazione = false;

    this.partitaIva = '';
    this.codice = macCod;
    this.centroPK = new CentroAziendale.PK(0, '');
    this.contatto = new Contatto();
    this.alimentazione = new Carburante();
    this.codice_stringa = '';
    this.costi = [];
    this.CUAA_Proprietario = '';
    this.data_Immatricolazione = AGRODATAINIZIO;
    this.data_Rilascio_Autorizzazione = new Date();
    this.data_Ultima_Revisione = new Date();
    this.data_Ultima_Manutenzione = new Date();
    this.Data_Carico = AGRODATAINIZIO;
    this.Data_Scarico = AGRODATAFINE;
    this.data_Ultima_Taratura = AGRODATAINIZIO;
    this.scadenza_Taratura = AGRODATAFINE;
    this.descrizione = '';
    this.dettaglio_1 = new MacchineDettaglio1();
    this.dettaglio_2 = new MacchineDettaglio2();
    this.ageaCod = new MacchineCodificaAgea();
    this.finalita = new FinalitaMacchina();
    this.marca = new DittaMacchina();
    this.modello = '';
    this.N_Autorizzazione_Trasporto = '';
    this.n_Immatricolazione = '';
    this.n_Immatricolazione_Rimorchio = '';
    this.note = '';
    this.potenza = '';
    this.proprietario = '';
    this.stato_Utilizzo = '';
    this.immagineGrande = undefined;
    this.immaginePiccola = undefined;
    this.VIN = '';
    this.BTM_Serial = '';
    this.ExternalAPIKey = '';
    this.HubIoT_PlatformDestination = new BaseCodeDescr(0);
    this.caratteristiche = [];
    this.gerarchiaPadre = undefined;
    this.gerarchiaFigli = [];
    this.taratura_Ugello = 0;
    this.numero_certificato = '';
    this.targa = '';
    this.telaio = '';
    this.tipo = new Macchine();
    this.tipo_Targa = new TipoTarga();
    this.titolo_Possesso = new TitoloDiPossesso(0);
    this.unita_Misura = new UnitaDiMisura(0, '');
    this.validita = new IntervalloTemporale(AGRODATAINIZIO, AGRODATAFINE);
    this.visibilitaPubblica = false;
    this.visibileControlloGestione = false;
  }

}

export class Caratteristica {
  codice: number;
  caratteristica: CaratteristicaDettagli;
  valore: string;
  validita_inizio: Date;
  validita_fine: Date;

  constructor(codice: number, Caratteristica_Cod: number, Caratteristica_Des: string, valore: string, validita: IntervalloTemporale) {
    this.codice = codice;
    this.caratteristica = new CaratteristicaDettagli(Caratteristica_Cod, Caratteristica_Des);
    this.valore = valore;
    this.validita_inizio = validita.inizio;
    this.validita_fine = validita.fine;
  }
}

export class CaratteristicaDettagli {
  codice: number;
  descrizione: string;

  constructor(codice: number, descrizione: string) {
    this.codice = codice;
    this.descrizione = descrizione;
  }
}

export class MacchinaGerarchia {
  ID: number;
  codice: number;
  macchina: ParcoMacchine;
  legame: BaseCodeDescr;
  desclegame: string;
  udm: UnitaDiMisura;
  qta: number;
  validita_inizio: Date;
  validita_fine: Date;

  constructor(ID: number, codice: number, Componente: ParcoMacchine, Legame_Cod: number, Legame_Des: string, DescrizioneLegame: string, UdM_Cod: number, UdM_Des: string, Qta: number, validita: IntervalloTemporale) {
    this.ID = ID;
    this.codice = codice;
    this.macchina = Componente;
    this.legame = new BaseCodeDescr(Legame_Cod, Legame_Des);
    this.desclegame = DescrizioneLegame;
    this.udm = new UnitaDiMisura(UdM_Cod, UdM_Des);
    this.qta = Qta;
    this.validita_inizio = validita.inizio;
    this.validita_fine = validita.fine;
  }
}

export class RateoTempo {
  
  Rateo_Cod: number;
  DataInizio: Date;
  DataFine: Date;
  OraInizio: Date;
  OraFine: Date;
  Rotazione: number
  ValiditaInizio: Date;
  ValiditaFine: Date;

  constructor(Rateo_Cod: number, DataInizio: Date, DataFine: Date, OraInizio: Date, OraFine: Date, Rotazione: number, ValiditaInizio: Date, ValiditaFine: Date) {
    this.Rateo_Cod = Rateo_Cod;
    this.DataInizio = DataInizio;
    this.DataFine = DataFine;
    this.OraInizio = OraInizio;
    this.OraFine = OraFine;
    this.Rotazione = Rotazione;
    this.ValiditaInizio = ValiditaInizio;
    this.ValiditaFine = ValiditaFine;
  }

}

export class LinkedMachine<T> extends ParcoMacchine {
  linkValidity: IntervalloTemporale;
  linkedItemPK: T;

  constructor(pk: T, macCod: number) {
    super(macCod);
    this.linkedItemPK = pk;
  }
}

export enum enum_TipoLegame {
  Gerarchia = 1
}
